src/data/markdown/translated-guides/en/07 Misc/05 k6 Extensions.md

```diff
@@ -1,4 +1,395 @@
+- Custom metric emission can be done by creating new metrics using
+  [`stats.New()`](https://github.com/grafana/k6/blob/v0.33.0/stats/stats.go#L449)
+  and emitting them using [`stats.PushIfNotDone()`](https://github.com/grafana/k6/blob/v0.33.0/stats/stats.go#L429).
+  For an example of this see the [`xk6-remote-write` extension](https://github.com/dgzlopes/xk6-remote-write).
+
+
+### Writing a new Output extension
+
+Output extensions are similarly written, but have a different API and performance
+considerations.
+
+The core of an Output extension is a struct that implements the [`output.Output`
+interface](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L57). For example:
+
+<CodeGroup labels={["log.go"]} lineNumbers={[false]}>
+
+```go
+package log
+
+import (
+	"fmt"
+	"io"
+
+	"go.k6.io/k6/output"
+	"go.k6.io/k6/stats"
+)
+
+// Register the extension on module initialization.
+func init() {
+	output.RegisterExtension("logger", New)
+}
+
+// Logger writes k6 metric samples to stdout.
+type Logger struct {
+	out io.Writer
+}
+
+// New returns a new instance of Logger.
+func New(params output.Params) (output.Output, error) {
+	return &Logger{params.StdOut}, nil
+}
+
+// Description returns a short human-readable description of the output.
+func (*Logger) Description() string {
+	return "logger"
+}
+
+// Start initializes any state needed for the output, establishes network
+// connections, etc.
+func (o *Logger) Start() error {
+	return nil
+}
+
+// AddMetricSamples receives metric samples from the k6 Engine as they're
+// emitted and prints them to stdout.
+func (l *Logger) AddMetricSamples(samples []stats.SampleContainer) {
+	for i := range samples {
+		all := samples[i].GetSamples()
+		for j := range all {
+			fmt.Fprintf(l.out, "%d %s: %f\n", all[j].Time.UnixNano(), all[j].Metric.Name, all[j].Value)
+		}
+	}
+}
+
+// Stop finalizes any tasks in progress, closes network connections, etc.
+func (*Logger) Stop() error {
+	return nil
+}
+```
+
+</CodeGroup>
+
+Notice a couple of things:
+
+- The module initializer `New()` receives an instance of
+  [`output.Params`](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L36).
+  With this object the extension can access the output-specific configuration,
+  interfaces to the filesystem, synchronized stdout and stderr, and more.
+- `AddMetricSamples` in this example simply writes to stdout. In a real-world
+  scenario this output might have to be buffered and flushed periodically to avoid
+  memory leaks. Below we'll discuss some helpers you can use for that.
+
+
+#### Additional features
+
+- Output structs can optionally implement additional interfaces that allows them to
+  [receive thresholds](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L79),
+  [test run status updates](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L94)
+  or [interrupt a test run](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L88).
```

Comment:
Should these link to pkg.go.dev or GitHub? Our API docs aren't great, so I thought GH would be better.

Reply:
I'm fine with GitHub, should we use permalinks?

Reply:
I am for pkg.go.dev. This links will get stale really fast. It's better to link to the docs and improve that over time than to just link to code. Also the docs links to the code so people can use that.

Reply:
Changed links to `pkg.go.dev` in 51e583a.